File: packages/client/src/live/live-stream.ts
import { createChatClient, type ChatClient } from '@ravenkash/chat';
import { createRTCClient, type Room, type RTCClient } from '@ravenkash/rtc';
import type { LiveStreamCredentials, LiveStreamRole } from './types';

/**
 * Raven Live Streaming, for the browser.
 *
 * Not a third real-time system. `LiveStream.join()` takes the credentials
 * your backend minted (`POST /v1/live-streams/:id/hosts` or
 * `.../viewer-tokens`) and gives you a real `@ravenkash/rtc` `Room`, plus a
 * real `@ravenkash/chat` `ChatClient` when the credentials include one.
 * The very same classes those packages already document, not wrappers
 * round them. `LiveStream.room` **is** an `@ravenkash/rtc` `Room`, and
 * every method and event on it behaves exactly as documented there.
 *
 * The "host publishes, viewers subscribe" behaviour of a live stream is
 * just ordinary `room.enableCamera()`/`enableMicrophone()` on one side and
 * nothing at all on the other. The server settled who gets which via the
 * RTC token's permission grant long before either browser saw a byte. This
 * class never asks you for a role and never sends one; `role` is purely
 * informational, read back from what your backend told it.
 *
 * ```ts
 * // Your own backend calls POST /v1/live-streams/:id/hosts (or
 * // .../viewer-tokens) and sends the response to this device.
 * const stream = await LiveStream.join(credentials);
 *
 * if (stream.isHost) {
 *   await stream.room.enableCamera();
 *   await stream.room.enableMicrophone();
 * }
 *
 * stream.room.on('trackSubscribed', (track) => videoEl.srcObject = track.mediaStream);
 * await stream.react('❤️');
 * await stream.leave();
 * ```
 */
export class LiveStream {
  readonly streamId: string;
  readonly role: LiveStreamRole;
  /** The underlying `@ravenkash/rtc` client. Go straight to it for anything `LiveStream` doesn't expose. */
  readonly rtc: RTCClient;
  /** The joined room: camera, microphone, screen share, participants, connection stats. Everything `@ravenkash/rtc`'s Room documents. */
  readonly room: Room;
  /** Present only when the credentials included a chat token. `undefined` for an RTC-only integration. */
  readonly chat?: ChatClient;

  private readonly chatRootMessageId?: string | null;

  private constructor(
    streamId: string,
    role: LiveStreamRole,
    rtc: RTCClient,
    room: Room,
    chat: ChatClient | undefined,
    chatRootMessageId: string | null | undefined,
  ) {
    this.streamId = streamId;
    this.role = role;
    this.rtc = rtc;
    this.room = room;
    this.chat = chat;
    this.chatRootMessageId = chatRootMessageId;
  }

  /** True for HOST and CO_HOST, the only roles the server grants publish permissions to. A VIEWER's `room` is always subscribe-only, and the server enforces that; this check doesn't. */
  get isHost(): boolean {
    return this.role === 'HOST' || this.role === 'CO_HOST';
  }

  /**
   * Joins a live stream. One entry point for hosts and viewers alike.
   * Which one you end up as follows entirely from the credentials your
   * backend minted; it's never a parameter here.
   */
  static async join(credentials: LiveStreamCredentials): Promise<LiveStream> {
    const rtc = createRTCClient({
      token: credentials.rtc.token,
      endpoint: credentials.rtc.endpoint,
      iceServers: credentials.rtc.iceServers,
      telemetryUrl: credentials.rtc.telemetryUrl,
    });
    const room = await rtc.join(credentials.streamId);

    let chat: ChatClient | undefined;
    if (credentials.chat) {
      chat = createChatClient({
        token: credentials.chat.token,
        apiUrl: credentials.chat.apiUrl,
        chatUrl: credentials.chat.chatUrl,
      });
      await chat.connect({ room: credentials.chat.conversations[0] });
    }

    return new LiveStream(credentials.streamId, credentials.role, rtc, room, chat, credentials.chatRootMessageId);
  }

  /**
   * The TikTok-style heart tap. Reactions ride on the stream's own root
   * chat message through `@ravenkash/chat`'s existing, already-aggregated
   * reaction model (`chat.messages.addReaction`). No second real-time
   * primitive invented specially for this. Every viewer tapping the same
   * emoji collapses into one count, exactly like reacting to any other
   * chat message.
   */
  async react(emoji: string): Promise<void> {
    if (!this.chat) {
      throw new Error(
        'This LiveStream has no chat credentials; react() needs the `chat` field on the credentials passed to join().',
      );
    }
    if (!this.chatRootMessageId) {
      throw new Error('This stream has no chatRootMessageId to react to.');
    }
    await this.chat.messages.addReaction(this.chatRootMessageId, emoji);
  }

  /**
   * Leaves the room and disconnects chat. The stream carries on for
   * everybody else; this only tears down *your* connection to it. Ending
   * it for everyone is a server-side action
   * (`POST /v1/live-streams/:id/end`), not something a client gets to do.
   */
  async leave(): Promise<void> {
    await this.rtc.leave();
    await this.chat?.disconnect();
  }
}

/** Functional alias, to match `createRTCClient`, `createChatClient` and `createRaven`. */
export function joinLiveStream(credentials: LiveStreamCredentials): Promise<LiveStream> {
  return LiveStream.join(credentials);
}

File: sdks/flutter/raven_live/lib/src/live_stream.dart
import 'package:raven_chat/raven_chat.dart'
    show RavenChat, RavenChatErrorCode, RavenChatException;
import 'package:raven_rtc/raven_rtc.dart' show Raven, RavenRoom;

import 'types.dart';

/// Raven Live Streaming on Flutter.
///
/// ```dart
/// final stream = await RavenLiveStream.join(credentials);
///
/// if (stream.isHost) {
///   await stream.room.enableCamera();
///   await stream.room.enableMicrophone();
/// }
///
/// await stream.react('❤️');
/// await stream.leave();
/// ```
///
/// A thin composition, not a parallel implementation: [room] *is* a
/// `raven_rtc` [RavenRoom] and [chat] *is* a `raven_chat` [RavenChat].
/// Every method and stream documented for those packages: participant
/// changes, camera/microphone, messages, reactions, typing, presence;
/// works unchanged on what this class hands back. Host vs. viewer publish
/// permission is enforced entirely server-side by the RTC token grant;
/// this class never sends a role, only reflects one it was given.
class RavenLiveStream {
  RavenLiveStream._(this.credentials, this._raven, this._chat, this._room);

  /// Joins a stream and connects its chat, if it has one.
  static Future<RavenLiveStream> join(
    RavenLiveStreamCredentials credentials,
  ) async {
    final raven = Raven(
      token: credentials.rtc.token,
      endpoint: credentials.rtc.endpoint,
      iceServers: credentials.rtc.iceServers,
    );

    RavenChat? chat;
    final chatCredentials = credentials.chat;
    if (chatCredentials != null) {
      chat = RavenChat(
        token: chatCredentials.token,
        apiUrl: chatCredentials.apiUrl,
        chatUrl: chatCredentials.chatUrl,
      );
    }

    final room = await raven.join(credentials.streamId);

    if (chat != null && chatCredentials != null) {
      final conversation = chatCredentials.conversations.isNotEmpty
          ? chatCredentials.conversations.first
          : credentials.streamId;
      try {
        await chat.connect(conversation);
      } catch (error) {
        // The room joined; don't strand it because chat failed. A caller
        // that genuinely needs chat will see `stream.chat` connect state
        // reflect the failure via its own connectionStateChanges stream.
        await room.leave();
        room.dispose();
        rethrow;
      }
    }

    return RavenLiveStream._(credentials, raven, chat, room);
  }

  final RavenLiveStreamCredentials credentials;
  final Raven _raven;
  final RavenChat? _chat;
  final RavenRoom _room;

  /// Which stream this is.
  String get streamId => credentials.streamId;

  /// This identity's role: never something inferred, only ever what the
  /// credentials said.
  RavenLiveStreamRole get role => credentials.role;

  /// `true` for [RavenLiveStreamRole.host] and [RavenLiveStreamRole.coHost].
  bool get isHost => role.isHost;

  /// The stream's room. Every `raven_rtc` API works on it unchanged.
  RavenRoom get room => _room;

  /// The stream's chat, if it has one attached. Every `raven_chat` API
  /// works on it unchanged.
  RavenChat? get chat => _chat;

  /// Reacts to the stream: the same reaction model web, CLI, and the
  /// server SDKs use, attached to the chat message every viewer's
  /// reaction lands on.
  ///
  /// Throws a [RavenChatException] with
  /// [RavenChatErrorCode.notInRoom] if this stream has no chat
  /// conversation attached.
  Future<void> react(String emoji) async {
    final chat = _chat;
    final rootMessageId = credentials.chatRootMessageId;
    if (chat == null || rootMessageId == null) {
      throw const RavenChatException(
        RavenChatErrorCode.notInRoom,
        'This stream has no chat conversation attached, so reactions are unavailable.',
      );
    }
    await chat.addReaction(rootMessageId, emoji);
  }

  /// Leaves the room and disposes the chat connection. Ending the stream
  /// itself (`LIVE` -> `ENDED`) is a separate, server-side call: see
  /// `@ravenkash/server`'s/`raven-sdk`'s `end_stream()`.
  Future<void> leave() async {
    await _raven.leave();
    _chat?.dispose();
  }
}
